src/graph.rs

```diff
@@ -27,27 +29,55 @@ impl GitGraph {
             repository,
             commits,
         };
-        graph.assign_branches(indices)?;
+        graph.assign_branches(indices, settings)?;
 
         Ok(graph)
     }
-    fn assign_branches(&mut self, indices: HashMap<Oid, usize>) -> Result<(), Error> {
-        let branches = self.repository.branches(None)?;
-        for branch in branches {
-            let (branch, branch_type) = branch?;
-            if branch_type == BranchType::Local {
-                let reference = branch.get();
-                if let Some(name) = reference.name() {
-                    if let Some(oid) = reference.target() {
-                        let idx = indices[&oid];
-                        self.commits[idx].branches.push(name[11..].to_string());
-                    }
+    fn assign_branches(
+        &mut self,
+        indices: HashMap<Oid, usize>,
+        settings: &Settings,
+    ) -> Result<(), Error> {
+        let branches_ordered = Self::branches_persistence_order(&self.repository, settings)?;
+
+        for branch in branches_ordered {
+            let reference = branch.get();
+            if let Some(name) = reference.name() {
+                if let Some(oid) = reference.target() {
+                    let idx = indices[&oid];
+                    self.commits[idx].branches.push(name[11..].to_string());
                 }
             }
         }
 
         Ok(())
     }
+
+    fn branches_persistence_order<'repo>(
+        repository: &'repo Repository,
+        settings: &Settings,
+    ) -> Result<Vec<Branch<'repo>>, Error> {
+        let branches = repository
+            .branches(Some(BranchType::Local))?
+            .filter_map(|bt| bt.ok().map(|bt| bt.0))
```

Comment:
It seems inconsistent to forward the errors return by `branches`, but to swallow the errors indicated by the individual elements.

Reply:
My idea here was that it really won't work if I can't get the branches at all, but to just leave out individual branches that produce an error, and continue. But honestly, I did not investigate what might cause individual branches to fail.

Reply:
> So you think it should then also fail completely?

It depends on whether this is recoverable and the information on the other branches can be trusted. Even then, emitting a warning seems appropriate. Failing completely when an underlying library reports an error is certainly correct. But of course, it can also reduce applicability. As a general rule, I would say failing early is preferable if the underlying issues are not well understood.

Reply:
Ah, trying to solve that I now also understand the `collect::<Result<Vec<_>, Error>>()?` from your suggestion [here](https://github.com/mlange-42/git-graph/pull/2#discussion_r544498028), that I did not get to work (due to `.ok()` of course).

Reply:
(In this particular situation, I suspect that this kind of error reporting exists just because the underlying C API needed to be adapted to Rust's iterator protocol where the only way to report a failure during iteration is to yield `Result`s which is generally considered inconvenient and the raison d'etre for crates like [`fallible-iterator`](https://docs.rs/fallible-iterator/0.2.0/fallible_iterator/). This would lend further support for failing early as the underlying Git errors are probably not as distinct as the Rust API on top of them would suggest. (More generally speaking, Rust's reification of effects using `Result`, `Option`, `Iterator` etc. shows a weak spot here as it does not compose nicely, c.f. https://without.boats/blog/the-problem-of-effects/)

Reply:
Ok, thanks for the link!